File: php/tests/YAMLStarTest.php
<?php

require_once __DIR__ . '/../src/YAMLStar.php';

use YAMLStar\YAMLStar;

function assertSameValue($expected, $actual, string $message): void
{
    if ($expected !== $actual) {
        fwrite(STDERR, "not ok - $message\n");
        fwrite(STDERR, 'expected: ' . var_export($expected, true) . "\n");
        fwrite(STDERR, 'actual:   ' . var_export($actual, true) . "\n");
        exit(1);
    }
    echo "ok - $message\n";
}

$ys = new YAMLStar();

assertSameValue(['key' => 'value'], $ys->load('key: value'), 'load mapping');
assertSameValue(['a', 'b'], $ys->load("- a\n- b\n"), 'load sequence');
assertSameValue(['doc1', ['a' => 1]], $ys->loadAll("---\ndoc1\n---\na: 1\n"), 'load all');
assertSameValue("key: value\n", $ys->dump(['key' => 'value']), 'dump mapping');
assertSameValue(
    ['items' => ['a', 'b'], 'flag' => true, 'text' => '42'],
    $ys->load($ys->dump(['items' => ['a', 'b'], 'flag' => true, 'text' => '42'])),
    'dump round trip'
);
assertSameValue("---\ndoc1\n---\na: 1\n", $ys->dumpAll(['doc1', ['a' => 1]]), 'dump all');

if (!preg_match('/^\d+\.\d+\.\d+(-SNAPSHOT)?$/', $ys->version())) {
    fwrite(STDERR, "not ok - version\n");
    exit(1);
}
echo "ok - version\n";

try {
    $ys->load('key: "unclosed');
    fwrite(STDERR, "not ok - malformed YAML should fail\n");
    exit(1);
} catch (RuntimeException $e) {
    if (!str_contains($e->getMessage(), 'libyamlstar:')) {
        throw $e;
    }
    echo "ok - malformed YAML error\n";
}

$ys->close();

File: ruby/test/yamlstar_test.rb
# frozen_string_literal: true

require "minitest/autorun"
require "yamlstar"

class YAMLStarTest < Minitest::Test
  def setup
    @ys = YAMLStar.new
  end

  def teardown
    @ys.close
  end

  def test_load
    assert_equal({"key" => "value"}, @ys.load("key: value"))
    assert_equal(["a", "b"], @ys.load("- a\n- b\n"))
  end

  def test_load_all
    assert_equal(["doc1", {"a" => 1}], @ys.load_all("---\ndoc1\n---\na: 1\n"))
  end

  def test_dump
    assert_equal("key: value\n", @ys.dump({"key" => "value"}))
    value = {"items" => ["a", "b"], "flag" => true, "text" => "42"}
    assert_equal(value, @ys.load(@ys.dump(value)))
  end

  def test_dump_all
    assert_equal("---\ndoc1\n---\na: 1\n", @ys.dump_all(["doc1", {"a" => 1}]))
  end

  def test_version
    assert_match(/\A\d+\.\d+\.\d+(-SNAPSHOT)?\z/, @ys.version)
  end

  def test_error
    error = assert_raises(YAMLStar::Error) { @ys.load('key: "unclosed') }
    assert_match(/libyamlstar:/, error.message)
  end
end
